infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/RestoreRedundancyMergeSegments.java

```diff
@@ -0,0 +1,54 @@
+package org.corfudb.infrastructure.orchestrator.actions;
+
+import com.google.common.collect.Sets;
+
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+
+import javax.annotation.Nonnull;
+
+import org.corfudb.infrastructure.orchestrator.Action;
+import org.corfudb.runtime.CorfuRuntime;
+import org.corfudb.runtime.exceptions.OutrankedException;
+import org.corfudb.runtime.view.Layout;
+
+/**
+ * This action attempts to restore redundancy for all servers across all segments
+ * starting from the oldest segment. It then collapses the segments once the set of
+ * servers in the 2 oldest subsequent segments are equal.
+ * Created by Zeeshan on 2019-02-06.
+ */
+public class RestoreRedundancyMergeSegments extends Action {
+
+    @Nonnull
+    @Override
+    public String getName() {
+        return "RestoreRedundancyAndMergeSegments";
+    }
+
+    @Override
+    public void impl(@Nonnull CorfuRuntime runtime) throws OutrankedException,
+            ExecutionException,
+            InterruptedException {
+
+        Layout layout;
+
+        // Catchup all servers across all segments.
+        do {
+            runtime.invalidateLayout();
+            layout = runtime.getLayoutView().getLayout();
+
+            // Currently the state is transferred for the complete segment.
+            // TODO: Add stripe specific transfer granularity for optimization.
+            // Get the set of servers present in the second segment but not in the first
+            // segment.
+            Set<String> lowRedundancyServers = Sets.difference(
```

Comment:
What is this trying to do ?

Reply:
This finds the difference in the 2 sets: Segment 1 and Segment 0 to find out nodes present in 1 and not in 0. These nodes need state transfer before the segments can be merged.